File: python/year_2019/day_01.py
"""
https://adventofcode.com/2019/day/1
"""

from itertools import repeat, takewhile
from pathlib import Path


def part_1(text):
    return sum(int(mass) // 3 - 2 for mass in text.split())


def part_2(text):
    def fuel(mass):
        fuels = (mass := mass // 3 - 2 for _ in repeat(None))
        return sum(takewhile(lambda mass: 0 < mass, fuels))

    return sum(map(fuel, map(int, text.split())))


def _read_input() -> str:
    return (Path(__file__).with_name("test_data") / "day_01.txt").read_text()


def test_part_1_input():
    assert part_1(_read_input()) == 3408471


def test_part_1_example_1():
    assert part_1("12") == 2


def test_part_1_example_2():
    assert part_1("14") == 2


def test_part_1_example_3():
    assert part_1("1969") == 654


def test_part_1_example_4():
    assert part_1("100756") == 33583


def test_part_2_input():
    assert part_2(_read_input()) == 5109803


def test_part_2_example_1():
    assert part_2("14") == 2


def test_part_2_example_2():
    assert part_2("1969") == 966


def test_part_2_example_3():
    assert part_2("100756") == 50346

File: rust/year_2019/src/day_01.rs
pub fn part_1(input: &str) -> u32 {
    parse_input(input).map(|mass| mass / 3 - 2).sum()
}

pub fn part_2(input: &str) -> u32 {
    parse_input(input)
        .map(|mut fuel| {
            let mut total = 0;
            while fuel >= 9 {
                fuel = fuel / 3 - 2;
                total += fuel;
            }
            total
        })
        .sum()
}

fn parse_input(input: &str) -> impl Iterator<Item = u32> {
    input.lines().map(|line| line.parse().unwrap())
}

#[cfg(test)]
mod test {
    use test_case::test_case;

    const INPUT: &str = include_str!("../test_data/day_01.txt");

    #[test_case("12" => 2)]
    #[test_case("14" => 2)]
    #[test_case("1969" => 654)]
    #[test_case("100756" => 33583)]
    #[test_case(INPUT => 3408471)]
    fn part_1(input: &str) -> u32 {
        super::part_1(input)
    }

    #[test_case("14" => 2)]
    #[test_case("1969" => 966)]
    #[test_case("100756" => 50346)]
    #[test_case(INPUT => 5109803)]
    fn part_2(input: &str) -> u32 {
        super::part_2(input)
    }
}
